Add more details to SDK MeterProvider and View (#1730)
open-telemetry/opentelemetry-specification: specification/metrics/sdk.md

matter whether a `Meter` was obtained from the `MeterProvider` before or after
the configuration change). Note: Implementation-wise, this could mean that
`Meter` instances have a reference to their `MeterProvider` and access
configuration only via this reference.

### Shutdown

TODO

### ForceFlush

TODO

### View

`View` gives the SDK users flexibility to customize the metrics they want. Here
are some examples when `View` is needed:

* Customize which [Instrument](./api.md#instrument) to be processed/ignored. For
  example, an [instrumented library](../glossary.md#instrumented-library) can
  provide both temperature and humidity, but the application developer only
  wants temperature information.
* Customize the aggregation - if the default aggregation associated with the
  Instrument does not meet the expectation. For example, an HTTP client library
  might expose HTTP client request duration as [Histogram](./api.md#histogram)
  by default, but the application developer only wants the total count of
  outgoing requests.
* Customize which attribute(s) to be reported as metrics dimension(s). For
  example, an HTTP server library might expose HTTP verb (e.g. GET, POST) and

>   HTTP status code (e.g. 200, 301, 404). The application developer might only

Reviewer: Not sure if this belongs, but adjusting the sampling of `Exemplar`s is *likely* an aggregator problem.  I can write up my thoughts here but I investigated what Prometheus has been up to in this space and I think likely exemplar sampling should be tied to aggregator.  If not, we should call it out as a thing users may want to change.

  care about HTTP status code (e.g. reporting the total count of HTTP requests
  for each HTTP status code). There are also extreme scenario that the
  application developer does not need any dimension (e.g. just get the total
  count of all incoming requests).
* Add additional dimension(s) from the [Context](../context/context.md). For
  example, a [Baggage](../baggage/api.md) value might be available indicating
  whether an HTTP request is coming from a bot/crawler or not. The application
  developer might want this to be converted to a dimension for HTTP server

>   metrics (e.g. the request/second from bots vs. real users).

Reviewer: FYI - In practice this likely is only achievable for synchronous instruments.

Author: I've added a note here.

  would be used by default. This will be used as the name of the [metrics
  stream](./datamodel.md#events--data-stream--timeseries).
* The Instrument selection criteria (required), which covers:
  * The `name` of the Meter (optional).
  * The `version` of the Meter (optional).
  * The `schema_url` of the Meter (optional).
  * The `name` of the Instrument (required).
  * Individual language client MAY choose to support more criteria.
* The configuration for the resulting [metrics
  stream](./datamodel.md#events--data-stream--timeseries):
  * The `description`. If not provided, the Instrument `description` would be
    used by default.
  * The `attribute keys` (optional). If not provided, all the attribute keys
    will be used by default (TODO: once the Hint API is available, the default
    behavior should respect the Hint if it is available).
  * The `extra dimensions` which come from Baggage/Context (optional). If not
    provided, no extra dimension will be used.
  * The `aggregation` (optional) to be used. If not provided, the default
    aggregation (based on the type of the Instrument) will be applied.

> If there are more than one Instruments meeting the selection criteria,

Reviewer 2: I thought about what to expect here, and came up with this: 

If a View matches more than one Instrument, there should one or more free variables among the listed criteria. If so, then I want the View to be instantiated once for every matching Instrument with free variables preserved. If there are no free variables, I figure, then the SDK must have broken some other part of the specification. Can we say simply state that the result is "unspecified", for now?

Author: Proposal: change the spec to treat this as error, and let SDK do proper error handling.

Two scenarios:
1. We have View("foobar"), later we discovered an Instrument("foobar")
2. We have multiple Views with the same name.

individual language implementation can decide what is the proper behavior, as
long as the behavior is well defined and deterministic.

If the `aggregation` is not provided, and the default aggregation doesn't have
sufficient information (e.g. a Histogram without any bucket configuration), the
SDK SHOULD provide a way to let the user know (e.g. expose [self-diagnostics
logs](../error-handling.md#self-diagnostics)).

> If there is no View registered, all the Instruments associated with the

Reviewer: I think this might need some rewording here.

Specifcally you may need to do some transition sentences to denote you're talking about set up of an SDK.

E.g.

"When initalizing an Instrument, if no View is registered, then a view with default aggregation is configured for this instrument"

The wording of this sentence and the previous are bit confusing / hard t interpret for me.

Author: This is entangled with the support for  multiple pipelines, which is a TODO.
I think we probably need to have a separate PR in order to fully resolve it.

We can discuss this in the upcoming Thu. Metrics SIG meeting.

MeterProvider SHOULD be used.

> If there is any View registered, only the registered View(s) SHOULD be used.

Reviewer: When/where?

> If there is any View registered, only the registered View(s) SHOULD be used.

Reviewer 3: If I'm understanding correctly, when an application developer wants to customize one Instrument and registers a View, they now have to explicitly opt-in to _all_ other Instruments from a MeterProvider.

Is there any concern about this developer experience? E.g., "I had all these metrics coming from various instrumentation libraries "for free," but by creating a single View I'm now required to create a View for every metric I care about."

Author: Yes, we've discussed about the usability part in the Metrics SIG meetings.

I think we need to tackle this in multiple steps:

1. We know that we need View (or whatever name we decided to call this) to allow application developers to do customization.
2. As the first step, we need to hammer out what are necessary things to enable the customization.
3. We might leave some room in the spec so later we can take the feedback and make it simpler, this can happen later and doesn't have to block the entire effort. The bottom line is that we don't introduce breaking changes.

Reviewer 4: +1 to "The usability is not very good, if by adding one 'View', I now need to opt-in to all other instruments". But this can solved as a follow up.

Reviewer 2: I think we can address the usability problem. I want to remove "If there are more than one Instruments meeting the selection criteria." above.

It sounds like a View with no fields configured, i.e., an empty view, will match all instruments. Could we specify that:

1. Views are ordered within a MeterProvider
2. Every instrument selects all matching Views in order subject to name uniqueness
3. Every instrument with no matching View has a basic view with the default aggregation applied

Users can:
1. Use a "match-all" View to get the default behavior for all unmatched instruments
2. Prioritize specific Views over generic Views
3. Use "blackhole" Views to match and discard metrics to bypass the default View.

I think this addresses the usability problem because, in the example scenario only two Views are needed: one to match a specific instrument, and one to do the default thing for all the other instruments.

Author: Let me update the PR based on @jmacd's suggestion.

  stream](./datamodel.md#events--data-stream--timeseries).
* The Instrument selection criteria (required), which covers:
  * The `name` of the Meter (optional).
  * The `version` of the Meter (optional).
  * The `schema_url` of the Meter (optional).
  * The `name` of the Instrument (required).
  * Individual language client MAY choose to support more criteria.
* The configuration for the resulting [metrics
  stream](./datamodel.md#events--data-stream--timeseries):
  * The `description`. If not provided, the Instrument `description` would be
    used by default.
  * The `attribute keys` (optional). If not provided, all the attribute keys
    will be used by default (TODO: once the Hint API is available, the default
    behavior should respect the Hint if it is available).
  * The `extra dimensions` which come from Baggage/Context (optional). If not
    provided, no extra dimension will be used. Please note that this only
    applies to [synchronous Instruments](./api.md#synchronous-instrument), any
    `extra dimensions` configured on [asynchronous
    Instruments](./api.md#asynchronous-instrument) will be considered as error.

>   * The `aggregation` (optional) to be used. If not provided, a default

Reviewer 4: "how" the aggregation to be used is specified is missing from this PR. I hope that'll be covered in a subsequent follow up. Some questions which i'd like to see if spec would cover:
1. How to specify aggregator? Is it something similar "Tracing Samplers", where one can pick one of the built-in samplers, or build own. (this'd require adding Aggregator as a defined thing. But I think it was already mentioned in last meetings that we'll come back to it later, based on actual need.) 
2. Would View determine (or have the ability to determine) if delta/cumulative temporality be used.

Not a blocking comment :)

Reviewer: For the JVM the first cut/prototype has this: https://github.com/jsuereth/opentelemetry-java/blob/wip-metrics-api/sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/view/ViewBuilder.java

Specifying an aggregator is done via the builder, so it's not an enum or a class (but could be both).  

In practice, this felt a little wonky when it comes to trying to output "Long" or "Double" values.   Is more information on this coming as part of the aggregator work?

Reviewer 2: Can we let Long vs. Double be a language-dependent decision, not a specification question?

Reviewer: To the extent OTLP exposes Long vs. Double, it matters.    At a minimum, I'd ask for some words in the spec that *allows* languages to expose Long vs.Double if they wish.   E.g. In java it's odd if a `LongCounter` instrument aggregates to a DoubleSum in OTLP when the default aggregation is LongSum.

For languages that don't care about types, we shouldn't force them to as long as there's no suprising oddities in output OTLP.  In this case, there *is* an oddity unless SDKs are given freedom to address this.

Reviewer 2: That would be an API spec question, then. I think we can resolve this comment.

## MeterProvider

### Meter Creation

New `Meter` instances are always created through a `MeterProvider` (see

> [API](./api.md#meterprovider)). The `name` and `version` arguments supplied to

Reviewer 2: ```suggestion
[API](./api.md#meterprovider)). The `name`, `version` (optional), and `schema_url` (optional) arguments supplied to
```

Author: Thanks! Updated.

> the `MeterProvider` must be used to create an

Reviewer: s/must/MUST/

Reviewer 5: Should we add the schemaUrl as well to this, since that's now been added to the APIs in general?

Reviewer 2: ```suggestion
the `MeterProvider` MUST be used to create an
```

Author: Fixed.

[`InstrumentationLibrary`](https://github.com/open-telemetry/oteps/blob/main/text/0083-component.md)
instance which is stored on the created `Meter`.

Configuration (i.e., [MeasurementProcessors](#measurementprocessor),
[MetricProcessors](#metricprocessor), [MetricExporters](#metricexporter) and

> [`Views`](#view)) MUST be managed solely by the `MeterProvider` and it MUST

Reviewer: nit:  Stronger verbage:
`it MUST allow configuration for all components implemented in the SDK.  This configuration MUST be available during creation or SDK initialization.`

Reviewer 2: I have thought about this in the opposite direction. First you assemble an SDK that has all the metrics features, settings, and parts assembled into a whole. Then you ask it to be a MeterProvider. 

(In the OTel-Go prototype, it's the "Controller" component that ultimately acts as MeterProvider. Managing configuration happened before the controller is constructed, in my thinking, and what construction does is prepare a configured MeterProvider.)

Author: > nit: Stronger verbage:
> `it MUST allow configuration for all components implemented in the SDK. This configuration MUST be available during creation or SDK initialization.`

I guess `configuration MUST be available during creation or SDK initialization` is way too strong. For example, the tracing spec gives the power to specific language clients if they want to support adding SpanProcessor/Exporter dynamically after the initialization ([trace sdk spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#tracer-creation) _"The TracerProvider MAY provide methods to update the configuration."_).

What I'm trying to do here, is to give the same flexibility rather than forcing all the configuration to be available during creation/initialization. In this case, if we feel it is hard to implement, we (I mean OpenTelemetry Java, Python, .NET, etc.) can choose to require all configurations to be provided ahead of time. Later if we figured out good usage scenario and better way to support dynamic configuration change, we won't get cornered/forbidden by the spec.

Author: Discussed during the SIG meeting, all the attendees think it is okay to mark it as resolved.
If anyone has concern, please reopen it.

provide some way to configure all of them that are implemented in the SDK, at
least when creating or initializing it.

> The `MeterProvider` MAY provide methods to update the configuration. If

Reviewer 4: Is the usage of "MAY" intentional here - for example, the current spec allows a language to allow no config updates after a MeterProvider is built. Is that allowed?

Reviewer 5: I would hope that an implementation can be immutable, as language preference would specify. If it MUST be updatable, we MUST call that out very explicitly.

Author: Yes it is intentional. It shares the spirit (and wording) with the [tracing spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#tracer-creation).

Certain language clients might prefer immutable pattern as @jkwatson mentioned above, I think it is very related to the language paradigm and performance consideration (e.g. certain SDK implementation might prefer to emit low level CPU instructions for the maximum performance based on the assumption that configuration will not be updated).

configuration is updated (e.g., adding a `MetricProcessor`), the updated
configuration MUST also apply to all already returned `Meters` (i.e. it MUST NOT
matter whether a `Meter` was obtained from the `MeterProvider` before or after
the configuration change). Note: Implementation-wise, this could mean that
`Meter` instances have a reference to their `MeterProvider` and access
configuration only via this reference.

### Shutdown

TODO

> ### ForceFlush

Author: I think ForceFlush would only make sense when push exporters exist. This is something we should clarify when we added details to this section later.

Reviewer: Isn't OTLP the *first* exporter we have and isn't that push?

I agree clarifying later is fine, but I'd like for this to be part of MVP (e.g. for the cloud function use case).

Reviewer 2: I think it's fine to have this mirror the BatchSpanProcessor, which has ForceFlush.

Reviewer 6: @reyang we discussed a bit of a different model, indeed for the metrics export pipeline this makes sense, but we talked that the SDK should expose a Collect functionality and be pull based.

Before defining all these components probably it makes sense to have an overall architecture diagram.

Author: I've added several diagrams to explain the overall flow.

TODO

### View

`View` gives the SDK users flexibility to customize the metrics they want. Here
are some examples when `View` is needed:

* Customize which [Instrument](./api.md#instrument) to be processed/ignored. For
  example, an [instrumented library](../glossary.md#instrumented-library) can
  provide both temperature and humidity, but the application developer only
  wants temperature information.
* Customize the aggregation - if the default aggregation associated with the
  Instrument does not meet the expectation. For example, an HTTP client library
  might expose HTTP client request duration as [Histogram](./api.md#histogram)
  by default, but the application developer only wants the total count (sum

>   aggregation) of outgoing requests.

Reviewer 7: In this example, wouldn't changing to a sum aggregation result in a sum of the durations observed not the total count of requests?

Reviewer 5: I agree. this should be a `count` aggregation, rather than `sum`.

* Customize which metrics dimension(s) to be reported. For example, an HTTP
  server library might expose HTTP verb (e.g. GET, POST), HTTP status code (e.g.

>   200, 301, 404), and a [Baggage](../baggage/api.md) value indicating whether

Reviewer 4: is there any reason why the example uses a dimension coming from Baggage?

Author: The metrics dimension(s) could come from:
* attributes that are provided to the instrument
* baggage
* context

So the SDK View API has to provide a way for the user to say "I need 3 dimensions, 2 of them should be coming from attributes X and Y, while the 3rd dimension Z should be coming from Baggage['foobar']".

Reviewer 4: It'd be easier if we split this into 2:
1. Library is reporting some dimensions, User wants to use View to select a subset of those.
2. Library is reporting some dimensions. User wants to use View to add more dimensions, which can come from baggage/context.

Customize which metrics dimension(s) to be reported. For example, an HTTP server
library might expose HTTP verb (e.g. GET, POST), HTTP status code (e.g. 200,
301, 404). The application developer might only care about HTTP status code
(e.g. reporting the total count of HTTP requests for each HTTP status code).
There are also extreme scenario that the application developer does not need any
dimension (e.g. just get the total count of all incoming requests).

Add additional dimension(s) to be reported. For example, an HTTP server
library might expose HTTP verb (e.g. GET, POST), HTTP status code (e.g. 200,
301, 404). The application developer wants to track an additional dimension "BotSource",
which is available in the Baggage (but not provided by the HttpServer library).
Views can be configured to include additional dimensions from context/baggage.

Author: Changed to two separate scenarios.

  the request is coming from a bot/crawler or not. The application developer
  might only care about HTTP status code (e.g. reporting the total count of HTTP
  requests for each HTTP status code). There are also extreme scenario that the
  application developer does not need any dimension (e.g. just get the total
  count of all incoming requests).

> The MeterProvider MUST provide a way to register Views, and here are the

Reviewer 4: Do we expect MeterProvider to allow adding/removing Views's on the fly? or all views must be registered when creating the MeterProvider, and no changes are allowed afterwards.

Reviewer 5: My vote would be to make the MeterProvider implementation immutable, but that also could be a language-specific implementation detail.

Author: This is covered here https://github.com/open-telemetry/opentelemetry-specification/pull/1730/files#r641860398.

In a nutshell, the SDK implementation has freedom to decide whether configuration is fully mutable (everything can be changed), partially mutable (e.g. processors can be added, but views cannot be added) or not at all.

Reviewer 8: Just a note, when implementing a prototype for Views in the [Python](https://github.com/open-telemetry/opentelemetry-python/pull/596) metrics SDK, it was quite difficult to account for the use cases in which the configuration of the MeterProvider changing during runtime, especially when it related to finding aggregation of metrics together that had the same keys (if Views changed during runtime, we had to reevaluate which "keys" from different metrics would be grouped together and aggregated together). I realize that this prototype was implemented long before, but the principal should be the same. Just something to think about.

Reviewer 2: I am wary of trying to implement dynamic re-configuration for metrics SDKs and support any language that lets it be an option for each language SIG to consider. I think this will be an interesting topic later in time!

Author: This is the same discussion https://github.com/open-telemetry/opentelemetry-specification/pull/1730#discussion_r641860398.

Please refer to my response https://github.com/open-telemetry/opentelemetry-specification/pull/1730#discussion_r641893375.

> The MeterProvider MUST provide a way to register Views, and here are the

Reviewer 9: Is there a concept of a "default" (each instrument kind has a default aggregation) view? And if I register a new view, what happens to this "default" view? Is it disabled under some condition?

If I assuming a View can select multiple instruments. What happens if multiple views wind up targeting the same instrument? How do we resolve conflicts (i.e. name collision, duplicate metrics, etc...)?

Reviewer 2: I would like this decision to be explicit, not implicit. Can you register a view and cancel its existing registration? If you're renaming the metric, that's probably a good choice, but you could also want to keep the original and add another aggregation by a new name. I think the user should decide.

Reviewer 5: Recommend rephrasing to "The SDK must provide the means to register Views with a MeterProvider" to account for immutable configuration MeterProvider implementations.

Author: I've updated the description based on our discussion during the 6/24/2021 SIG meeting.

inputs:

> * The Instrument selection criteria (required), which MUST cover:

Reviewer: From a  'raw level SDK' this seems fine, but as specified i think you NEED a helper method on top of this.

E.g.  Can the user just glob-match on Version + Kind to get measurements based on Meter/Instrument-name?

I'm worried the way this language reads right now.   I think it makes sense to tease apart "What SDK implementors of metric processors need" and "What SDK Users defining views need".   does it make sense to split these two?

Effectively for users, your'e giving them a 'query' language to select metrics.
For the SDK, you're defining a data model "index" that needs to allow rapid processing.

i agree the 4 things you list are what belong on the index.

Reviewer 4: This is missing something to select "Aggregation". "Customize the aggregation" is a listed use-case.

Reviewer: I don't think this portion of the API needs to specify aggregation.  Instead we *do* need to clarify if any selected instrument has its default aggregation removed, or if it still exists.   It's unclear what would happen if I were, e.g. to do a "Select all" here for my metrics.   Do they all get removed?

  * The `name` of the Meter.
  * The `version` of the Meter.
  * The `name` of the Instrument.

>   * The `kind` of the Instrument.

Reviewer 8: Curious, what is the purpose of providing the `kind` of the instrument, if instruments are uniquely identified only by the `name` and the `meter` it belongs to?

Author: The initial thinking was that filter can by something like "give me all the instruments (regardless of the name) that are histograms".

Author: Based on the discussion from 6/22/2021 Metrics SIG meeting, we will reduce the scope by only allowing one exact match. This means we don't need `kind` now.

> * The `name` of the View (optional). If not provided, the Instrument `name`

Author: For discussion: if multiple instruments were picked by the "selection criteria", do we want to give them the same View name?

Potential approaches:
1. We only allow name and description to be changed if the selection criteria returned 0 or 1 instrument.
2. We leave the behavior as UNDEFINED and let each SDK handle that.
3. We allow multiple instruments to be selected and replaced with the same name (I could imagine some scenario but it is definitely NOT a common case).

Reviewer 5: what is the purpose of the View `name`? How is it used?

Author: For example, an HTTP server library has an instrument:

```python
http_server_duration = meter.create_histogram(
    name="http.server.duration",
    description="measures the duration of the inbound HTTP request",
    unit="milliseconds",
    value_type=float)

http_server_duration.Record(100, http_method="GET", http_scheme="http"})
```

The application developer is saying "I don't care about the distribution (or my backend doesn't support histogram), all I need is the sum of durations across all finished HTTP requests", then the developer can use the View API to find the instrument "http.server.duration", change the aggregation to Sum, and change the name to "http.server.duration.experimental".

Reviewer 5: ah, so view name becomes the resulting instrument name for the purposes of export? Interesting. How about units? Can views change the units of the resulting data?

Reviewer 10: > ah, so view name becomes the resulting instrument name for the purposes of export?

this is not apparent from the text, needs to be stated explicitly that this is the purpose of the view name.

Reviewer 9: It would be good to also clarify what happens if different Views result in the same name.

Reviewer: I think we should be "crisp" in language here:

1. The `View` defines the `Metric` name (I.e. the OTLP stream)
2. By Default, `Instrument` name becomes a `Metric` name when using the default aggregation.

+1 to not allowing views of different "metric types" of the same name (similar too the instrument restriction).

Author: I've updated the wording, trying to find the best balance here. PTAL.

  would be used by default.
* The `description` of the View (optional). If not provided, the Instrument
  `description` would be used by default.

> * The `attribute keys` (optional). If not provided, all the attribute keys will

Reviewer 9: If I configure my view to want keys B and C.

1. Do I fill in measurements missing B and/or C with some defaults?
2. Only match if measurements contain both B and C?
3. Somehow configurable for each B and/or C?

Reviewer 9: I'll interpret this to mean "these are the keys I'll want to keep, and drop all other keys".

> * The `attribute keys` (optional). If not provided, all the attribute keys will

Author: Do we think it is worthy to call out explicitly that "SDK MUST allow people to specify 'I don't need any attribute keys'" or it is okay to imply it?

Reviewer 5: I think if they are listed as options here, then it's implied that the SDK has to allow them to be specified.

>   be used by default (TODO: once the Hint API is available, the default behavior

Reviewer: I'm not understanding how the Hint API interacts here, given this is a query-filter (effectively) against measurments.  AFAIK - the Hint API only controls the aggregation, not the inputs-to-aggregation.

Reviewer 2: (Hint is no longer mentioned.)

## MeterProvider

### Meter Creation

New `Meter` instances are always created through a `MeterProvider` (see
[API](./api.md#meterprovider)). The `name`, `version` (optional), and
`schema_url` (optional) arguments supplied to the `MeterProvider` MUST be used
to create an
[`InstrumentationLibrary`](https://github.com/open-telemetry/oteps/blob/main/text/0083-component.md)
instance which is stored on the created `Meter`.

Configuration (i.e., [MeasurementProcessors](#measurementprocessor),
[MetricProcessors](#metricprocessor), [MetricExporters](#metricexporter) and

> [`Views`](#view)) MUST be managed solely by the `MeterProvider` and it MUST
> provide some way to configure all of them that are implemented in the SDK, at
> least when creating or initializing it.

Reviewer 5: ```suggestion
[`Views`](#view)) MUST be managed solely by the `MeterProvider` and the SDK MUST
provide a way to configure all options that are implemented by the SDK. This MAY be done at the
time of MeterProvider creation if appropriate.
```

The `MeterProvider` MAY provide methods to update the configuration. If
configuration is updated (e.g., adding a `MetricProcessor`), the updated
configuration MUST also apply to all already returned `Meters` (i.e. it MUST NOT
matter whether a `Meter` was obtained from the `MeterProvider` before or after
the configuration change). Note: Implementation-wise, this could mean that
`Meter` instances have a reference to their `MeterProvider` and access
configuration only via this reference.

### Shutdown

TODO

### ForceFlush

TODO

### View

> `View` gives the SDK users flexibility to customize the metrics they want. Here
> are some examples when `View` is needed:

Reviewer 5: ```suggestion
A `View` provides SDK users with the flexibility to customize the metrics that are output by the SDK. Here
are some examples when a `View` might be needed:
```

Author: Updated, thanks!

> * Customize which [Instrument](./api.md#instrument) to be processed/ignored. For

Reviewer 5: ```suggestion
* Customize which [Instruments](./api.md#instrument) are to be processed/ignored. For
```

Author: Updated, thanks!

  example, an [instrumented library](../glossary.md#instrumented-library) can

>   provide both temperature and humidity, but the application developer only
>   wants temperature information.

Reviewer 5: ```suggestion
  provide both temperature and humidity, but the application developer might only
  want the temperature.
```

Author: Updated, thanks!

* Customize the aggregation - if the default aggregation associated with the

>   Instrument does not meet the expectation. For example, an HTTP client library

Reviewer 5: ```suggestion
  Instrument does not meet the needs of the user. For example, an HTTP client library
```

Author: Updated, thanks!

  might expose HTTP client request duration as [Histogram](./api.md#histogram)

>   by default, but the application developer only wants the total count of

Reviewer 5: ```suggestion
  by default, but the application developer might only want the total count of
```

Author: Updated, thanks!

  outgoing requests.

> * Customize which attribute(s) to be reported as metrics dimension(s). For

Reviewer 5: ```suggestion
* Customize which attribute(s) are to be reported as metrics dimension(s). For
```

Author: Fixed, thanks!

  example, an HTTP server library might expose HTTP verb (e.g. GET, POST) and
  HTTP status code (e.g. 200, 301, 404). The application developer might only
  care about HTTP status code (e.g. reporting the total count of HTTP requests

>   for each HTTP status code). There are also extreme scenario that the
>   application developer does not need any dimension (e.g. just get the total

Reviewer 5: ```suggestion
  for each HTTP status code). There could also exist extreme scenarios in which the
  application developer does not need any dimension (e.g. just get the total
```

Author: Updated, thanks!

  count of all incoming requests).
* Add additional dimension(s) from the [Context](../context/context.md). For
  example, a [Baggage](../baggage/api.md) value might be available indicating
  whether an HTTP request is coming from a bot/crawler or not. The application
  developer might want this to be converted to a dimension for HTTP server
  metrics (e.g. the request/second from bots vs. real users).

The SDK MUST provide the means to register Views with a `MeterProvider`. Here
are the inputs:

* The Instrument selection criteria (required), which covers:

>   * The `name` of the Instrument(s), with wildcard support (required).

Reviewer 5: i think we need the kind of instrument here, as well.

Author: We have a statement which allows language clients to add more (e.g. point type, kind of the instrument, etc.). Do you think that's sufficient?

I wish to avoid making "kind of instrument" an "optional input" which we require all languages to implement.

Reviewer 5: It may be optional, but I think this will be an extremely common option, and hence I think we should specify it explicitly, and require all languages to support it.

Reviewer 2: I take it that this is to allow users to say which instrument they expect when configuring the view, meaning to prevent the rule from applying to the wrong kind of data.

Author: Please vote which option you would prefer:

@jkwatson and I discussed offline and we have different perspectives:
* @jkwatson believes that instrument type is a very important selection criteria, and SHOULD be implemented by all clients (as an optional parameter). For example, select all the Histograms and aggregate as min/max/sum/count.
* @reyang believes that instrument name is the most common criteria, and most users would use naming convention. For example, select all the "http.*.duration" and set the buckets to [x,y,z...].

We have two options:
* add `instrument type` as an optional criteria.
* leave it out for now, individual language can still do it since the SDK spec gives that flexibility _**individual language client MAY choose to support more criteria**_.

Author: Conclusion during the SIG:
1. add instrument type as a required criteria (optional).
2. change name to be optional.
3. mention that if all the optional criteria are missing, treat it as a fault. We encourage SDK to blow up, but leave space if people don't want to be aggressive (language might have some general preference), and point to the error handling doc.

Author: Clarify that the criteria are additive.

Author: I've made the updates, please take another look.

  * The `name` of the Meter (optional).
  * The `version` of the Meter (optional).
  * The `schema_url` of the Meter (optional).
  * Individual language client MAY choose to support more criteria. For example,
    a strong typed language MAY support point type (e.g. allow the users to
    select Instruments based on whether the underlying type is integer or
    double).
* The `name` of the View (optional). If not provided, the Instrument `name`
  would be used by default. This will be used as the name of the [metrics
  stream](./datamodel.md#events--data-stream--timeseries).
* The configuration for the resulting [metrics
  stream](./datamodel.md#events--data-stream--timeseries):
  * The `description`. If not provided, the Instrument `description` would be
    used by default.
  * A list of `attribute keys` (optional). If not provided, all the attribute
    keys will be used by default (TODO: once the Hint API is available, the
    default behavior should respect the Hint if it is available).
  * The `extra dimensions` which come from Baggage/Context (optional). If not
    provided, no extra dimension will be used. Please note that this only

>     applies to [synchronous Instruments](./api.md#synchronous-instrument).

Reviewer 5: I wonder if we should leave these out of the first pass. Otherwise, we need to figure out some standard way to figure out how to define the source of the information declaratively.

Reviewer 2: What do you mean by "define the source of the information"?

Author: I think having this would help us to move forward (once we have the skeleton of the spec, it is easier for others to come and contribute the "meat") 😄

Reviewer 5: I mean... how do you extract it from the context? Are you providing some sort of function on the context that would give you an attribute?

Reviewer 2: My understanding is that the SDK via its MeasurementProcessor has access to the context for this purpose.

Reviewer 5: sure...but how do you define a view to do that work? how do you specify how to extract data from the context?

  * The `aggregation` (optional) to be used. If not provided, a default
    aggregation will be applied by the SDK. The default aggregation is a TODO
    (e.g. first see if there is an explicitly specified aggregation from the
    View, if not, see if there is a Hint, if not, fallback to the default
    aggregation that is associated with the Instrument type), and for histogram,
    the default buckets would be (-&infin;, 0], (0, 5.0], (5.0, 10.0], (10.0,
    25.0], (25.0, 50.0], (50.0, 75.0], (75.0, 100.0], (100.0, 250.0], (250.0,

>     500.0], (500.0, 1000.0], (1000.0, +&infin;).

Reviewer 5: I think I would leave this out at this point, and just leave it as a TODO.

Author: Let's keep the buckets but add a comment clarifying that this is still under discussion and will change?

Reviewer 5: I'm mostly confused about what this is getting at. Are you trying to define how to do view selection? Or define default aggregations? I'm mostly just confused by this parenthetical paragraph, and don't think it helps clarify things. If you want to have a placeholder for default aggregations, I would put it in a different section, rather than putting it here parenthetically.

Author: I think this entire paragraph after TODO is something we'll change and move to a different place. And I think as long as it is a clear "TODO", we want to keep the information here so we don't have to start from scratch (e.g. spending another two weeks discussing about the default histogram buckets) once we make progress on default aggregation / configuration.

Author: Discussed during the Metrics SIG, I've updated the PR and removed this section for now. I will work with @victlu to capture this in a separate PR.

> The SDK SHOULD use the following logic to determine how to process an
> Instrument:

Reviewer 5: ```suggestion
The SDK SHOULD use the following logic to determine how to process recordings make with an 
Instrument:
```

Author: I'll change "recordings" to "Measurements" to align with the terminology, sounds good?

> * Determine the `MeterProvider` which "owns" the Instrument.
> * If the `MeterProvider` has no `View` registered, take the Instrument and apply
>     the default configuration.
> * If the `MeterProvider` has one or more `View`(s) registered, for each View:
>   * If the Instrument could match the instrument selection criteria:
>     * Try to apply the View configuration. If there is an error (e.g. the View
>       asks for extra dimensions from the Baggage, but the Instrument is
>       [asynchronous](./api.md#asynchronous-instrument) which doesn't have
>       Context) or a conflict (e.g. the View requires to export the metrics using
>       a certain name, but the name is already used by another View), provide a
>       way to let the user know (e.g. expose [self-diagnostics
>       logs](../error-handling.md#self-diagnostics)).
> * END.

Reviewer 2: Should it be possible to configure the behavior for an instrument only when none of configured Views applies?  
```suggestion
* Determine the `MeterProvider` which "owns" the Instrument.
* If the `MeterProvider` has no `View` registered, take the Instrument and apply
    the default configuration.
* If the `MeterProvider` has one or more `View`(s) registered, for each View:
  * If the Instrument could match the instrument selection criteria:
    * Try to apply the View configuration. If there is an error (e.g. the View
      asks for extra dimensions from the Baggage, but the Instrument is
      [asynchronous](./api.md#asynchronous-instrument) which doesn't have
      Context) or a conflict (e.g. the View requires to export the metrics using
      a certain name, but the name is already used by another View), provide a
      way to let the user know (e.g. expose [self-diagnostics
      logs](../error-handling.md#self-diagnostics)).
  * If the Instrument matches none of the configured `View`(s) and an optional
     default is provided, apply the default configuration.
* END.
```

Author: I think this is already supported if people provide a wildcard rule at the end.

For example:

```python
meterProvider
    .add_view('X', attributes=['x', 'y']) # instrument X will be selected, and mapped to metrics with dimensions 'x' and 'y'
    # anything else will be selected, with dimension 'x'
    # note that X will not selected here because it will create a conflict (duplicated name)
    .add_view('*', attributes=['x'])
```

Reviewer 2: I was looking at a slight difference. A wildcard rule by itself has no way to say "Apply this configuration when no other view applies".

Author: This has been resolved, please refer to [this comment](https://github.com/open-telemetry/opentelemetry-specification/pull/1730#discussion_r674182701)

Here are some examples:

```python
# Python
'''
+------------------+
| MeterProvider    |
|   Meter A        |
|     Counter X    |
|     Histogram Y  |
|   Meter B        |
|     Gauge Z      |
+------------------+
'''

# metrics from X and Y (reported as Foo and Bar) will be exported
meter_provider
    .add_view("X")
    .add_view("Foo", instrument_name="Y")
    .add_view(
        "Bar",
        instrument_name="Y",
        aggregation=HistogramAggregation(buckets=[5.0, 10.0, 25.0, 50.0, 100.0]))
    .set_exporter(PrometheusExporter())
```

```python
# all the metrics will be exported using the default configuration
meter_provider.set_exporter(ConsoleExporter())
```

```python
# all the metrics will be exported using the default configuration
meter_provider

>     .add_view("*") # a wildcard view that matches everything

Reviewer 5: I don't think this should be a valid call. There is a selector, but no actual configuration. I think that every view needs to have 2 components: 1) what instrument(s) are being selected and 2) what should be done to them.

Reviewer 5: And, I would suggest that the selector and the actual actions to be taken should be separated into two different parameters, rather than just stringing all of the options into a bunch of different parameters, willy-nilly. We should strive to make an API here that is really obvious and simple for users to use and understand.

Author: This is to catch everything by default. Here are the scenarios:

1. no view - you get everything with the default configuration
2. `add_view("X", configurations...)` - you get `X` with the custom configuration, and **nothing else**
3. `add_view("X", configurations...)` and `add_view("*")` - you get `X` with the custom configuration, and everything else with the default configuration

Reviewer 5: I feel very strongly that if you want to disable metrics by default, that should be an explicit API call on the SDK, rather than have it be implicit in adding a single view configuration.  [based on offline discussion with Reilly]

Reviewer: +1 to this.

Reviewer 2: I also agree. I proposed a slight change here that would give us two configuration points, https://github.com/open-telemetry/opentelemetry-specification/pull/1730#discussion_r673505159, one would be all the Views that are in effect, and the second would be the default settings that are in effect. I am trying to establish that View configuration should be separate from Default configuration.

Author: Discussed during the SIG and here goes the conclusion:

1. meterprovider to have an API supporting "disable defaults", how the API should be done is unspecified and up to language clients.
2. otel official sdks will be the same, the sdk spec will say something like "by default we'll get everything".

Author: I've updated the wording, please take another look.

[MetricProcessors](#metricprocessor), [MetricExporters](#metricexporter) and
[`Views`](#view)) MUST be managed solely by the `MeterProvider` and it MUST
provide some way to configure all of them that are implemented in the SDK, at
least when creating or initializing it.

The `MeterProvider` MAY provide methods to update the configuration. If
configuration is updated (e.g., adding a `MetricProcessor`), the updated
configuration MUST also apply to all already returned `Meters` (i.e. it MUST NOT
matter whether a `Meter` was obtained from the `MeterProvider` before or after
the configuration change). Note: Implementation-wise, this could mean that
`Meter` instances have a reference to their `MeterProvider` and access
configuration only via this reference.

### Shutdown

TODO

### ForceFlush

TODO

### Pipeline

A pipeline is a logical group of:

* An ordered list of [View](#view)s
* A [MeasurementProcessor](#measurementprocessor)
* An ordered list of [MetricProcessor](#metricprocessor)s
* A [MetricExporter](#metricexporter)

```text
               +----------------------+  +-----------------+
               |                      |  |                 |

> Measurements --> MeasurementProcessor +--> In-memory state +--+

Reviewer 6: I think this can be simplified by having the SDK support only one "export endpoint". If a need to support multiple pipelines is identified then we can have a Fanout API implementation which can redirect measurements to multiple SDK instances.

The SDK can become something like this:

```text
               +----------------------+  +-----------------+  +---------------+
               |                      |  |                 |  |               |
Measurements --> MeasurementProcessor +--> In-memory state +--+ MetricsReader +
               |                      |  |                 |  |               |
               +----------------------+  +-----------------+  +---------------+
```

In this case the proposed "SDK" is equivalent with 1 pipeline in the current proposal, but it does separation of concerns:
* Fanout is a property of an API implementation that does only this.
* SDK knows how to aggregate measurements by applying the View configurations and exposes the result.

`MetricsReader` - is a simple interface that exposes an API to collect already aggregated metrics via a `Collect` method (which may accept a filter if needed).

On the export "pipeline"/package we can have an `IntervalMetricsReader` that is capable of reading metrics from multiple producers and concatenates them to support Push like exporters. The `IntervalMetricsReader` can call into a `MetricsProcessor` if needed.

A prometheus (pull based exporter) can be implemented in two ways:
1. Accepts a list of MetricsProducers and simply queries them, sends them to the "processors" and then exposes them on the given port. In case of multiple exporters support this can be implemented using the Fanout API implementation + configuring multiple SDKs.
2. Implement a "StorageMetricsProcessor" which is capable of storing the results (for gauge applies last value, for delta sums adds the deltas, for cumulative keeps the last value, etc.) that is called by the `IntervalMetricsReader`.

The details about the pull based system can be defined in the followup PR that will define the export path.

Reviewer: I talked to @reyang about some of my own concerns around Pipeline.

Things I was not aware of:
- During a metrics SIG there was consensus around allowing *differing* view implementations per pipeline/exporter.  While I find this surprising, it's something to account for.
- I think the current way the specification is written doesn't make control flow obvious, but if i understand correctly from meeting w/ Reiley, he *does* envision of `MetricReader` style interface for the export channel.  I think formalizing this is a great idea.

Fundamentally, I think the issue comes down to whether or not we want to support MULTIPLE in-memory storages in one SDK + whether or not having different view configurations should be an SDK concern.

I like Bogdan's proposal from the standpoint that we have evidence it worked in the past, and it keeps the architecture somewhat simple.   I think we can even combine it with some of the API ideas outlined here in a follow CL.

Fundamentally, one of my concerns in the current SDK specifciatoin is it is almost forcing a `Measurement -> Queue -> Router -N-> Storages` arcthitecture.   In the current Java implementation (and the prototype API/SDK upgrade), we instead attempt to go `Measurement -> Storage` as efficiently as we can (Note: This is a deviation from how OpenCensus worked, which Bogdan explained was to address latency/speed concerns w/ OpenCensus).

To the extent we can keep the specification focus on *semantics* and *interfaces* rather than architecture, I'd like to do so.  I'm fully on board with the mechanism of specifying views.  I like the idea of simplifying pipeline configuration for the API.   I'm unsold (but will go with consensus) on views differing across exporters.   I'd prefer if the multi-pipeline aspect of this PR were in a different PR where we could hash out that discussion.

Reviewer: I talked to @reyang about some of my own concerns around Pipeline.

Things I was not aware of:
- During a metrics SIG there was consensus around allowing *differing* view implementations per pipeline/exporter.  While I find this surprising, it's something to account for.
- I think the current way the specification is written doesn't make control flow obvious, but if i understand correctly from meeting w/ Reiley, he *does* envision of `MetricReader` style interface for the export channel.  I think formalizing this is a great idea.

Fundamentally, I think the issue comes down to whether or not we want to support MULTIPLE in-memory storages in one SDK + whether or not having different view configurations should be an SDK concern.

I like Bogdan's proposal from the standpoint that we have evidence it worked in the past, and it keeps the architecture somewhat simple.   I think we can even combine it with some of the API ideas outlined here in a follow CL.

Fundamentally, one of my concerns in the current SDK specifciatoin is it is almost forcing a `Measurement -> Queue -> Router -N-> Storages` arcthitecture.   In the current Java implementation (and the prototype API/SDK upgrade), we instead attempt to go `Measurement -> Storage` as efficiently as we can (Note: This is a deviation from how OpenCensus worked, which Bogdan explained was to address latency/speed concerns w/ OpenCensus).

To the extent we can keep the specification focus on *semantics* and *interfaces* rather than architecture, I'd like to do so.  I'm fully on board with the mechanism of specifying views.  I like the idea of simplifying pipeline configuration for the API.   I'm unsold (but will go with consensus) on views differing across exporters.   I'd prefer if the multi-pipeline aspect of this PR were in a different PR where we could hash out that discussion.

Reviewer: I talked to @reyang about some of my own concerns around Pipeline.

Things I was not aware of:
- During a metrics SIG there was consensus around allowing *differing* view implementations per pipeline/exporter.  While I find this surprising, it's something to account for.
- I think the current way the specification is written doesn't make control flow obvious, but if i understand correctly from meeting w/ Reiley, he *does* envision of `MetricReader` style interface for the export channel.  I think formalizing this is a great idea.

Fundamentally, I think the issue comes down to whether or not we want to support MULTIPLE in-memory storages in one SDK + whether or not having different view configurations should be an SDK concern.

I like Bogdan's proposal from the standpoint that we have evidence it worked in the past, and it keeps the architecture somewhat simple.   I think we can even combine it with some of the API ideas outlined here in a follow CL.

Fundamentally, one of my concerns in the current SDK specifciatoin is it is almost forcing a `Measurement -> Queue -> Router -N-> Storages` arcthitecture.   In the current Java implementation (and the prototype API/SDK upgrade), we instead attempt to go `Measurement -> Storage` as efficiently as we can (Note: This is a deviation from how OpenCensus worked, which Bogdan explained was to address latency/speed concerns w/ OpenCensus).

To the extent we can keep the specification focus on *semantics* and *interfaces* rather than architecture, I'd like to do so.  I'm fully on board with the mechanism of specifying views.  I like the idea of simplifying pipeline configuration for the API.   I'm unsold (but will go with consensus) on views differing across exporters.   I'd prefer if the multi-pipeline aspect of this PR were in a different PR where we could hash out that discussion.

Author: I've removed the pipeline part from this PR based on our discussion during the SIG meeting. Will make it a separate PR.

               |                      |  |                 |  |
               +----------------------+  +-----------------+  |
                                                              |
  +-----------------------------------------------------------+
  |
  |  +-------------------+             +-------------------+ 
  |  |                   |             |                   | 
  +--> MetricProcessor 1 +--> ... ... -> MetricProcessor N +--+
     |                   |             |                   |  |
     +-------------------+             +-------------------+  |
                                                              |
  +-----------------------------------------------------------+
  |
  |  +----------------+
  |  |                |
  +--> MetricExporter +--> Another process
     |                |
     +----------------+
```

> `MeterProvider` MUST support multiple pipelines:

Reviewer 6: What is a pipeline based on the previous image?

Reviewer 6: Just saw the comment, it is fine.

```text
+------------------+
| MeterProvider    |
|   Meter A        +-----> Pipeline 1
|     Counter X    |
|     Histogram Y  +-----> ...
|   Meter B        |       ...
|     Gauge Z      |
|     ...          +-----> Pipeline N
|   ...            |
+------------------+
```

Each pipeline has an ordered list of [View](#view)s, which provides
configuration for:

* Whether the Measurements from a certain Instrument should go through the
  pipeline or not.
* How should the Measurements turn into Metrics.

> ### View

Reviewer 6: Would be good to map the View on the diagram at 76, and understand which components does it control/configure.

Author: Based on the discussion in the SIG meeting, I've scoped down the PR by removing pipelines.

  whether an HTTP request is coming from a bot/crawler or not. The application
  developer might want this to be converted to a dimension for HTTP server
  metrics (e.g. the request/second from bots vs. real users).

The SDK MUST provide the means to register Views with a [MeterProvider
Pipeline](#pipeline). Here are the inputs:

* The Instrument selection criteria (required), which covers:
  * The `name` of the Instrument(s), with wildcard support (required).
  * The `name` of the Meter (optional).
  * The `version` of the Meter (optional).
  * The `schema_url` of the Meter (optional).
  * Individual language client MAY choose to support more criteria. For example,
    a strong typed language MAY support point type (e.g. allow the users to
    select Instruments based on whether the underlying type is integer or
    double).
* The `name` of the View (optional). If not provided, the Instrument `name`
  would be used by default. This will be used as the name of the [metrics
  stream](./datamodel.md#events--data-stream--timeseries).
* The configuration for the resulting [metrics
  stream](./datamodel.md#events--data-stream--timeseries):

>   * The `description`. If not provided, the Instrument `description` would be

Reviewer: Should a view also be able to change "units" if it changes aggregation?

Reviewer 2: 👍

Author: From @jsuereth: Java has an aggregator that counts the number of recordings (so the unit from the Instrument might be "ms" duration, and the output from the View will be "number of requests").

This is really depending on "do we allow custom aggregators".

Author: @bogdandrutu doesn't think unit should be there at this moment, the suggestion is that we don't do anything for now.

Author: @reyang to create an issue to track this, treat as a non-blocking comment for this PR.

Author: Issue created #1822

    used by default.

>   * The `attribute keys` (optional). If not provided, all the attribute keys

Reviewer: What is the "type" of this value.  A list of attribute key strings?  A Regex?

Reviewer 2: Trying to keep it simple: can we start with just a list of key strings?

Author: I've updated it to "a list of `attribute keys`".

    will be used by default (TODO: once the Hint API is available, the default
    behavior should respect the Hint if it is available).
  * The `extra dimensions` which come from Baggage/Context (optional). If not
    provided, no extra dimension will be used. Please note that this only
    applies to [synchronous Instruments](./api.md#synchronous-instrument), any
    `extra dimensions` configured on [asynchronous

>     Instruments](./api.md#asynchronous-instrument) will be considered as error.

Reviewer: FYI - regarding attribute key usage in the JVM prototype, I did the following: https://github.com/jsuereth/opentelemetry-java/blob/wip-metrics-api/sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/view/AttributesProcessors.java

There's an `AttributesProcessor` interface that's locked down and provides basic primitives and "join" operations to construct your behavior around altering attributes.    Would this meet the specification?

Also what does "considered an error" mean in terms of how to handle this scenario?  May want to link to error-handling-in-the-sdk section.

Author: In general, the metrics SDK spec would want to give some room to the language clients so they can shoot for higher performance. @reyang to follow up with @jsuereth to change the wording in the spec to give that room.

Author: Filter attributes from the Measurements.
Take extra attributes from Baggage/Context.
Add extra attributes (most likely to be static per view, e.g. "aggregation=MAX").

Author: The error handling part is covered in the next section, I've removed the "... will be considered as error" part here to avoid duplication.

  * The `aggregation` (optional) to be used. If not provided, a default
    aggregation will be applied by the SDK. The default aggregation is a TODO
    (e.g. first see if there is an explicitly specified aggregation from the
    View, if not, see if there is a Hint, if not, fallback to the default
    aggregation that is associated with the Instrument type), and for histogram,

>     the default buckets would be [0, 5.0], (5.0, 10.0], (10.0, 25.0], (25.0,

Reviewer: Note:  Otel starts its histograms from `-Inf`, so your first two buckets should be  `(-Inf, 0],  (0, 5.0]`, not just `[0, 5.0]`

Reviewer 2: (On a tangent and just to be clear: Prometheus uses 0.005, 0.01, 0.05, ... through 10, and we have specified a factor of 1000 greater. Are we also recommending the use of Milliseconds for timing instruments, so the default histogram boundaries match the units?)

Author: These are good callouts. I think we need to track this in a separate PR, probably I should mark this part as TODO.

> Are we also recommending the use of Milliseconds for timing instruments, so the default histogram boundaries match the units?)

Excellent question @jmacd! I struggled a lot with this, and my answer is not simple.

Currently the semantic conventions (we all know they are still Experimental) are using milliseconds. For example https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/http-metrics.md#http-client. I think our default histogram has to align with semantic convention (otherwise it becomes less useful or even misleading).

I think using Seconds as the default will give precision/rounding issue if the target being measured are at CPU level (e.g. measured in nanoseconds or ticks) and we have to convert it to float/double. Note that this is less likely to impact Histogram since we don't sum things up, however it does affect semantic convention and counters.

Ideally I would want to avoid the SDK spec to have ANY bias/preference towards unit (we already said in the API spec that unit is opaque).

Author: Discussed this during the 7/13/2021 Metrics SIG meeting:
1. We've already discussed whether we want default histogram buckets or not, and the conclusion was YES.
2. Regarding which default should we choose, we discussed and concluded that we want to go with Prometheus default buckets, with a slight modification, that we'll use 5, 10, 25, 50 rather than 0.005, 0.01, 0.025, 0.05, etc. (because in OpenTelemetry we prefer milliseconds than seconds).

Author: I've added the `(-Inf, 0], (0, 5.0]` buckets based on the feedback from @jsuereth.

    50.0], (50.0, 75.0], (75.0, 100.0], (100.0, 250.0], (250.0, 500.0], (500.0,
    1000.0], (1000.0, +&infin;).

The SDK SHOULD use the following logic to determine how to process an
Instrument:

* Determine the MeterProvider which "owns" the Instrument.
* If the MeterProvider has no [Pipeline](#pipeline) registered, go to the END.
* For each Pipeline:
  * If the Pipeline has no `View` registered, take the Instrument and apply the
    default configuration.

>   * If the Pipeline has one or more `View`(s) registered, for each View:

Reviewer: Interesting, I was sharing views across all pipelines.   In this instance, it implies that storage *may* need to be isolated per-pipeline, as the interaction between `View` (and specifically `Aggregator`) with storage might be an implementation detail.

A few questions:

- Do all pipelines share the same MeasurementProcessor?
- Do you expect different internal storage of "Metrics" or "Accumlulations" per-pipeline given this architecture?
- How does this logic line up with the "late-binding" notion of instruments allowed in the API?   e.g. for cases where instrumentation grabs an instrument PRIOR to SDK being fully set up, and then trying to backfill the instrument to flow to the configured SDK.   Note: That's an "optional" choice for API/SDKs, but I think it has poor interaction here.
- Do pipelines have to be registered all at once, or can they be added later?

Reviewer 5: IMO, I think pipelines should be 100% independent of each other (at least, if the user wants to share a processor instance, or whatever across pipelines, that's their decision, not enforced by the SDK).
And, I think all pipelines should be fully configured & registered at MeterProvider creation time.

Reviewer 2: If there are three layers in a pipeline, Measurement processing -> Metric processing -> Exporting, it seems that the question is whether we can configure multiple pipelines that share common prefixes, for efficiency. As far as the mechanism goes, we just need to support "Multi-" interfaces for each: Multi-measurement processor, multi-metric processor, and multi-exporter. These components would pass individual metric events, accumulations, and aggregations to multiple of the downstream components. 

Now, the question seems to be how complicated a configuration language you want to recommend across all OTel SDKs, for users to set up optimized pipelines. I'd like to keep it simple. It would be nice if the user could configure 100% independent pipelines and, if you care to be fancy, the SDK can optimize that configuration into an efficient implementation by sharing pipeline prefixes. Let's not specify how that's done, in other words.

Author: > * Do all pipelines share the same MeasurementProcessor?

Conceptually no. The SDK can be smart and optimize things if the optimization yields the same outcome from the user perspective.

> * Do you expect different internal storage of "Metrics" or "Accumlulations" per-pipeline given this architecture?

No - from the spec perspective. Whether it is shared or isolated, is an implementation detail.
Yes - from the implementation perspective, I expect some language clients will choose to combine/share the in-memory state for sake of memory efficiency and perf.

> * How does this logic line up with the "late-binding" notion of instruments allowed in the API?> * Do pipelines have to be registered all at once, or can they be added later?

For most of the language clients I would expect what @jkwatson mentioned (https://github.com/open-telemetry/opentelemetry-specification/pull/1730#discussion_r668217764) - Pipelines are finalized at MeterProvider creation time. However individual implementation can support adding pipelines dynamically if they want, and the spec should allow that. We already covered it in this PR `The MeterProvider MAY provide methods to update the configuration`, and it is aligned with the [tracing sdk spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#tracer-creation) as well.

Reviewer: I'm on board with this.  Will have to adjust the Java implementation appropriately.

    * If the Instrument could match the instrument selection criteria:
      * Try to apply the View configuration. If there is an error (e.g. the View
        asks for extra dimensions from the Baggage, but the Instrument is
        [asynchronous](./api.md#asynchronous-instrument) which doesn't have
        Context) or a conflict (e.g. the View requires to export the metrics
        using a certain name, but the name is already used by another View),
        provide a way to let the user know (e.g. expose [self-diagnostics
        logs](../error-handling.md#self-diagnostics)).

>       * Stop processing the remaining Views (proceed to the next Pipeline).

Reviewer 4: If we stop processing remaining Views, it blocks the use-case of a single instrument being part of multiple views.
One example use case:
For a Counter with 2 attributes, A and B. By default (in the absence of views), all attributes are used for the timeseries. So the max # of timeseries possible is A*B.

If I write 2 views for the same instrument, one taking just "A" and other taking just "B" as attributes, we can have A+B max timeseries, as opposed to A*B. This approach requires a single instrument being part of more than one View.

Author: Nice catch! 💯

* END.

Here are some examples:

```python
# Python
'''
+------------------+
| MeterProvider    |
|   Meter A        |
|     Counter X    |
|     Histogram Y  |
|   Meter B        |
|     Gauge Z      |
+------------------+
'''

meter_provider.start_pipeline(
    # all the metrics will be exported using the default configuration
    pipeline: pipeline
        .set_exporter(ConsoleExporter())
).start_pipeline(
    # metrics from X and Y (reported as Foo and Bar) will be exported to Prometheus upon scraping
    pipeline: pipeline
        .add_view("X")

>         .add_view("Foo", instrument_name="Y")

Reviewer 11: What is `"Foo"`? [Here](https://github.com/open-telemetry/opentelemetry-specification/pull/1730/files#diff-32d5f289beed96900fa1e6e69a3ee9cc0b6503ce8883672011c44bc772c0fa2aR153) it says that the instrument name is required.

Reviewer 5: I think this is the name of the view, replacing the name of instrument "Y" in the resulting metrics.

        .add_view(
            "Bar",
            instrument_name="Y",
            aggregation=HistogramAggregation(buckets=[5.0, 10.0, 25.0, 50.0, 100.0]))
        .set_exporter(PrometheusExporter())
)
```

```python
meter_provider.start_pipeline(
    # all the metrics will be exported using the default configuration
    pipeline: pipeline
        .add_view("*") # a wildcard view that matches everything
        .set_exporter(ConsoleExporter())
)
```

```python
meter_provider.start_pipeline(
    # Counter X will be exported as cumulative sum
    pipeline: pipeline

>         .add_view("X", aggregation=SumAggregation(CUMULATIVE))

Reviewer 4: I think the end-user experience is bit complex if we expect them to pick the right temporality for the aggregation. Can we restrict view to only allow picking which type of Aggregation (SUM vs HISTOGRAM vs GAUGE vs SUMMARY),? Isn't the choice of using Delta vs Cumulative, a decision to be made by the Exporter alone?

Reviewer: I agree with the sentiment that if we can have users ONLY specify Sum/Histogram that'd be ideal.  However, OTLP e.g. supports both delta + cumulative.   Should the exporter force only one as a possibility?

Maybe we can get the best of both worlds if:

1. Aggregator assumes a default aggregation temporality (recommend cumulative)
2. Specifying aggregation temporality is an optional parameter.

Should this discussion be part of the aggreagtor specification in #1804 ?

Reviewer 5: specifying aggregation temporality should absolutely be optional. Exporters should definitely have access to the Views API and be able to tell the SDK that they *need* a given temporality for a given aggregation type, but I don't know if that should be the same mechanism or not.

Author: Short answer - the code example is trying to show what is possible, it is not suggesting that we should/must do it. And it should be addressed in another PR (probably #1804).

Long answer:
1. If we export to a place that only supports cumulative, we should convert delta to cumulative without asking the user to explicitly ask us.
2. If we export to a place that only supports delta, we should convert cumulative to delta without asking the user to explicitly ask us.
3. If we export to a place that supports both cumulative and delta (e.g. OTLP is a great example), we should have a default behavior so the user doesn't have to explicitly specify the temporality. However we should allow the user to specify what they want, and whatever the user specified, should be respected.
4. If the user specified something that is impossible (e.g. user explicitly specified DELTA while the exporter is Prometheus), we should treat it as error and report it properly.

The MeterProvider MUST provide a way to register Views, and here are the
inputs:

* The `name` of the View (optional). If not provided, the Instrument `name`
  would be used by default.
* The Instrument selection criteria (required), which covers:
  * The `name` of the Meter (optional).
  * The `version` of the Meter (optional).
  * The `schema_url` of the Meter (optional).
  * The `name` of the Instrument (required).
* The configuration for the resulting [metrics
  stream](./datamodel.md#events--data-stream--timeseries):
  * The `description`. If not provided, the Instrument `description` would be
    used by default.
  * The `attribute keys` (optional). If not provided, all the attribute keys
    will be used by default (TODO: once the Hint API is available, the default
    behavior should respect the Hint if it is available).
  * The `extra dimensions` which come from Baggage/Context (optional). If not
    provided, no extra dimension will be used.
  * The `aggregation` (optional) to be used. If not provided, the default
    aggregation (based on the type of the Instrument) will be applied.

If there is no Instrument meeting the selection criteria, the SDK SHOULD fail

> fast and let the caller know.

Reviewer 5: how can this be? The SDK will be configured with views, potentially before an API user has requested an instrument be created at all?

Author: I think there has to be a moment where the SDK has a complete understanding of all the meters and instruments for a given provider, and that's the point the SDK should validate the configuration (including view, processor, exporter).

Reviewer 5: I don't think that's true at all. Instruments can be created at any point in the lifecycle of the SDK, simply by the user making a request of the API's MeterProvider to get an instrument that has never previously been used.

Am I missing something here?

Author: > I don't think that's true at all. Instruments can be created at any point in the lifecycle of the SDK, simply by the user making a request of the API's MeterProvider to get an instrument that has never previously been used.

I agree with your statement regarding the API part. Whether we support that in the SDK or not can be another decision.

For example, I think an application can dynamically load other components and these components might create Instruments at any point of time (even during onexit if they are crazy enough). However in the SDK we can say, there is a certain point where we finalize everything, so the instruments created after that point will NOT be picked up by the finalized pipeline, and if the SDK supports adding new pipeline on-the-fly, the new pipeline can pick up the instruments that were created before that invocation time.

I think either way would work and there are many pros and cons (I think the Collector will go with the fully dynamic approach, the SDK might do it differently).

Reviewer 5: I have never heard anyone speak of a "finalized" metrics pipeline up until your statement here. We certainly don't have anything like that on the tracing side, where you are free to create named tracers whenever you please. 

Is this something that I just missed somewhere?

Author: I think tracing doesn't have such problem/challenge.

Take the following scenario for example:

1. The application is running, with the global MeterProvider taking data from an UpDownCounter named "foobar", exporting to Prometheus. The SDK is configured to take instrument(s) named "foobar" and export using the default setup (default name, description, attributes, aggregation).
2. After few minutes, another component got dynamically loaded, and it retrieved a different Meter from the global MeterProvider, and registered a Histogram using the name "foobar", without providing any bucket hint/configuration.
3. We have the following options:
  * Ignore Histogram("foobar"), write some internal error log - this might work, has its own pros/cons.
  * Interpret the Histogram("foobar") as UpDownCounter - this doesn't seem to be logical?
  * Ignore Histogram("foobar"), take the "finalized/snapshot" approach - this might work, has its own pros/cons.

Reviewer 5: Yes, having two instruments with the same name & different types is a problem, but I don't think that matters for this particular discussion. What if they didn't have the same name? Then, everything is just fine, isn't it? We just apply whatever hint + default + view applies to that instrument and go. 

Let's not confuse the "register two different instruments with the same name" issue with whether people can register instruments late in general.

Author: We've spent 40 minutes on this topic during the 6/24/2021 SIG meeting and @jkwatson feedback was taken.

I've updated the wording, PTAL.

individual language implementation can decide what is the proper behavior, as
long as the behavior is well defined and deterministic.

If the `aggregation` is not provided, and the default aggregation doesn't have
sufficient information (e.g. a Histogram without any bucket configuration), the

> SDK SHOULD fail fast and let the caller know.

Reviewer 5: I thought we were going to provide basic default aggregations for all the instruments. Has this changed?

Author: Not changed. Basic default aggregation will be provided for all instruments. However I think Histogram is a special case, we don't have "default buckets".

I can see three options:
1. downgrade Histogram to something else (e.g. only provide Min/Max)
2. force Histogram to have default buckets
3. report error if no buckets info found for Histogram (from Hint or View)

Author: Discussed during the 6/24/2021 SIG meeting:
1. We don't have default histogram buckets. The buckets have to be coming from either the View or the Hint.
2. If there is no histogram buckets, we will drop the Histogram, report error.
3. When we report the error, we MUST NOT fail the caller (e.g. throw exception, terminate the process). We can write self-diagnostics log by default, and the SDK could choose to expose a flag or callback which allows the user to change the behavior.

I've updated the wording, PTAL.

Author: Prometheus provides default buckets based on the HTTP duration.
The proposal from the SIG meeting - provide default bucket assuming that 90% of the histogram would be RPC duration.

Do we need to understand unit? That'll be hard, most likely we will try to optimize based on what the semantic convention recommended - e.g. treat everything as ms, even if the instrument provided has different unit (e.g. ms, ns, s) or type (int, double).

Prometheus has some default buckets - starting from 5ms to 10s.

[jmacd] if we have a dedicated timing instrument we might be able to avoid this problem (since time/duration and unit are implied by the timing instrument).

Conclusion: we treat all histogram values as if they were milliseconds, and we provide the default buckets based on Prometheus buckets.

[bogdan] regarding the actual wording, we might want to avoid "milliseconds" entirely, and say something like "for integer histogram, the histogram buckets would be [x, y, z, ...] and for double histogram the default buckets would be [a, b, c...]".

Note: later we might want to avoid doing something smarter (e.g. by understanding some units such as `second` or `nanosecond`), with the understanding that these ARE breaking changes to the spec/sdk.
